### PR TITLE
Allow filtering CKAN datastore on key/value or query string.

### PR DESCRIPTION
**Pull request by srowlands**
Allow for restricting result set returned from CKAN by:
- Query string (full text search)
- Filters (dictionary of key/value pairs, e.g: {"Postal_Code": "4066", "Age", "18"}

Makes the module more useful on large datasets.

### CODE REVIEW

File: modules/govcms_ckan_display/plugins/visualisation/bar_chart.inc [hunk 1]
```diff
@@ -27,7 +27,23 @@ $plugin = array(
 function govcms_ckan_display_bar_chart_view($file, $display, $config) {
   $element = array();
   $chart_class = 'ckan-bar-chart';
-  $response = govcms_ckan_client_request_records($file->resource_id);
+
+  // Apply filters to the record request.
+  if (!empty($config['ckan_filters']['q'])) {
```
Comment (jeremy-doghouse) [inline modules/govcms_ckan_display/plugins/visualisation/bar_chart.inc]: If the defaults are added to the plugin settings (at the top of this file) both the $search and $filters checks can be removed and the $config keys will always be present. Same with line & spline plugins (and maybe even the config form).

I know I shouldn't be looking :P but I can't help it, and don't want my baby getting full of unnecessarily IF statements :)

Comment (srowlands) [inline modules/govcms_ckan_display/plugins/visualisation/bar_chart.inc]: Cool, good call. Will add to defaults, that was a dumb oversight :+1:

File: modules/govcms_ckan_media/includes/govcms_ckan_media.field_config.inc [hunk 2]
```diff
@@ -279,6 +279,44 @@ function govcms_ckan_media_visualisation_default_key_config($form, $form_state,
     $options['all'][$field->id] = $field->id;
   }
 
+  // Optional filter and query parameters.
+  $form['ckan_filters'] = array(
+    '#type' => 'fieldset',
```
Comment (gollyg) [inline modules/govcms_ckan_media/includes/govcms_ckan_media.field_config.inc]: Translation?

File: modules/govcms_ckan_media/includes/govcms_ckan_media.field_config.inc [hunk 3]
```diff
@@ -279,6 +279,44 @@ function govcms_ckan_media_visualisation_default_key_config($form, $form_state,
...
+  if (isset($config['ckan_filters']['search'])) {
+    $ckan_query_default = $config['ckan_filters']['search'];
+  }
+  else {
+    $ckan_query_default = '';
+  }
+
+  $form['ckan_filters']['search'] = array(
+    '#type' => 'textfield',
+    '#title' => 'Full text query',
```
Comment (gollyg) [inline modules/govcms_ckan_media/includes/govcms_ckan_media.field_config.inc]: Translation?

File: modules/govcms_ckan_media/includes/govcms_ckan_media.field_config.inc [hunk 4]
```diff
@@ -279,6 +279,44 @@ function govcms_ckan_media_visualisation_default_key_config($form, $form_state,
...
+  if (isset($config['ckan_filters']['filters'])) {
+    $ckan_filters_default = $config['ckan_filters']['filters'];
+  }
+  else {
+    $ckan_filters_default = '';
+  }
+
+  $form['ckan_filters']['filters'] = array(
+    '#type' => 'textfield',
+    '#title' => 'Filters',
```
Comment (gollyg) [inline modules/govcms_ckan_media/includes/govcms_ckan_media.field_config.inc]: Translation?

**Comment by srowlands** [comment]
@gollyg resolved missing t()

File: modules/govcms_ckan_display/plugins/visualisation/bar_chart.inc [hunk 5]
```diff
@@ -27,7 +31,7 @@ $plugin = array(
 function govcms_ckan_display_bar_chart_view($file, $display, $config) {
   $element = array();
   $chart_class = 'ckan-bar-chart';
-  $response = govcms_ckan_client_request_records($file->resource_id);
+  $response = govcms_ckan_client_request_records($file->resource_id, $config['ckan_filters']['search'], $config['ckan_filters']['filters']);
```
Comment (gollyg) [inline modules/govcms_ckan_display/plugins/visualisation/bar_chart.inc]: No biggie, but maybe consider pulling these into variables for readability.

**Pull request merged**